### DISCUSSION TITLE
Avoiding null values in conditional output of array elements?

### DISCUSSION THREAD

**Discussion by digitalfiends** [schibsted/jslt#280] [Q&A] [answered]
This question is along the lines of my other one but I can't quite seem to get the behaviour that I'm looking for when trying to copy values from source JSON to an array value in the target.

The source has three name keys with string values that need to be conditionally copied to an array in the target.
The copy rules are:

1) if RC-GIVEN-NAME is present, copy to source to target array otherwise copy RC-T1-GIVEN-NAME and RC-T12ND-GIVEN-NAME to target array
2) it's also possible that none of the names fields exist in the source and therefore PersonGivenName should not be present in target.

Basically, the target array should either contain [ RC-GIVEN-NAME ] or [ RC-T1-GIVEN-NAME, RC-T12ND-GIVEN-NAME ] - if none of those fields exist, PersonGivenName in the output should be omitted.  If any of the source values are null, the target array should **not** contain a null element (which is what I'm trying to solve right now). 

Current JSLT transform - note this does not address requirement (2) above, only (1):

```
{
    "Client": {

        "PersonName": {
            "PersonGivenName": [
                fallback(.RC-INPUT-DTL-REC.RC-GIVEN-NAME, .RC-INPUT-DTL-REC.RC-T1-GIVEN-NAME),
                if(not(.RC-INPUT-DTL-REC.RC-GIVEN-NAME) and .RC-INPUT-DTL-REC.RC-T12ND-GIVEN-NAME)
                  .RC-INPUT-DTL-REC.RC-T12ND-GIVEN-NAME
            ]
       }
    }
}
```

This seems to work, except if T12ND-GIVEN-NAME is _not_ present in the source, the output array contains a null:

**Source JSON example: RC-GIVEN-NAME not present, RC-T12ND-GIVEN-NAME not present**

```
{
    "RC-INPUT-DTL-REC" : {
      "RC-T1-GIVEN-NAME" : "T1-GIVEN-NAME",
    }
}
```

Expected output:

```
{
  "Client" : {
    "PersonName" : {
      "PersonGivenName" : [ "T1-GIVEN-NAME" ]
    }
  }
}
```

Actual output:

```
{
  "Client" : {
    "PersonName" : {
      "PersonGivenName" : [ "T1-GIVEN-NAME", **null** ]
    }
  }
}
```

The 2nd expression ```if(not(.RC-INPUT-DTL-REC.RC-GIVEN-NAME) and .RC-INPUT-DTL-REC.RC-T12ND-GIVEN-NAME)``` is evaluating to null, resulting in a null being written to the target array.  Is it possible to get rid of that null value?

Thank you!

**UPDATE:**

@larsga solution works for what I need and this is what I changed my transform to:

```
{
    "Client": {
        "PersonName": {
            "PersonGivenName": 
                 [for ([
                         if(.RC-INPUT-DTL-REC.RC-GIVEN-NAME)
                           .RC-INPUT-DTL-REC.RC-GIVEN-NAME
                         else 
                           .RC-INPUT-DTL-REC.RC-T1-GIVEN-NAME, .RC-INPUT-DTL-REC.RC-T12ND-GIVEN-NAME
                       ]) . if (.)]
        }
    }
}

**Comment by larsga** [comment, marked as answer]
The easiest way to filter an array is to do

```
[for (<array>) . if (.)]
```

So you can do

```
{
    "Client": {

        "PersonName": {
            "PersonGivenName": [for ([
                fallback(.RC-INPUT-DTL-REC.RC-GIVEN-NAME, .RC-INPUT-DTL-REC.RC-T1-GIVEN-NAME),
                if(not(.RC-INPUT-DTL-REC.RC-GIVEN-NAME) and .RC-INPUT-DTL-REC.RC-T12ND-GIVEN-NAME)
                  .RC-INPUT-DTL-REC.RC-T12ND-GIVEN-NAME
            ]) . if (.)]
       }
    }
}
```

Or you can declare a variable to make it a little easier to see what's going on.

**Reply by digitalfiends** [reply]
Oh, I didn't think of that. I shall give that a try!  Thank you.

**Reply by ivanogor** [reply]
god bless you!
